src/collectives/device/all_to_all.h

```diff
@@ -32,29 +40,48 @@ class ncclFunction<ncclFuncAllToAll, ALGO, PROTO, FUNC, T, UNROLL> {
       const int nChunks = scklAlgo->nChunks;
       // assume that size is divisible by nchunks
       const ssize_t sizePerChunk = size/nChunks;
+      // sckl flags all start out with 0. this is used as a part of the flag to make sure different work items deal with different synchronization flags
+      // this still needs more work. when we make a way around the queue, the flag might have been set to undesired values. will be fixed in subsequent versions.
+      const int workIndex = args->index+1;
+      volatile struct scklFlag* scklFlags = comm->scklFlags;
       // Compute pointers
-      const T * __restrict__ thisInput = (const T*)args->sendbuff;
-      T * __restrict__ thisOutput = (T*)args->recvbuff;
+      T * thisInput = (T*)args->sendbuff;
+      T * thisOutput = (T*)args->recvbuff;
       int myRank = channel->ring.devUserRanks[0];
       int m1 = -1;
       int recvPeer = (sckltb->type == SCKL_RECV) ? sckltb->peer : m1;
       int sendPeer = (sckltb->type == SCKL_SEND) ? sckltb->peer : m1;
 
       ncclPrimitives<UNROLL, ALLTOALL_CHUNKSTEPS/ALLTOALL_SLICESTEPS, ALLTOALL_SLICESTEPS, T, 1, 1, 1, FUNC>
         prims(tid, nthreads, &recvPeer, &sendPeer, thisOutput, stepSize, channel, comm, ncclShmem->ptrs, 0);
-
-      for (ssize_t gridOffset = 0; gridOffset < sizePerChunk; gridOffset += loopSize) {
+      for (ssize_t gridOffset = 0, iter = 0; gridOffset < sizePerChunk; gridOffset += loopSize, iter++) {
         int realChunkSize = min(chunkSize, DIVUP(sizePerChunk-gridOffset,nChannels));
         ALIGN_SIZE(realChunkSize, nthreads*sizeof(uint64_t)/sizeof(T));
         ssize_t chunkOffset = gridOffset + channelId*realChunkSize;
         ssize_t offset;
         int nelem = min(realChunkSize, sizePerChunk-chunkOffset);
         for (int i = 0; i < sckltb->nsteps; i++){
-          offset = chunkOffset + sckltb->transfers[i] * sizePerChunk;
+          struct scklTransfer* sckltran = &sckltb->transfers[i];
+          if (sckltran->offset == -1) continue;
+          offset = chunkOffset + sckltran->offset * sizePerChunk;
+          T* thisbuffer = (sckltran->buffer == SCKL_INPUT_BUFFER) ? thisInput : thisOutput;
           if (sckltb->type == SCKL_SEND){
-            prims.directSend(thisInput + offset, offset, nelem);
+            int8_t dependentBid = sckltran->dependentRbid + scklNumBlocksPerChannel * channelId;
+            int8_t dependentStep = sckltran->dependentStep;
+            if (sckltran->dependentRbid >= 0){
+              if (tid == 0){
+                uint64_t goalFlag = COMPUTE_FLAG(workIndex, iter, dependentStep);
+                while ((scklFlags + dependentBid)->flag < goalFlag){};
```

Comment:
I am still a little wary about this spin loop. Flag definitely has to be volatile here.